x/stake/keeper_test.go

```diff
@@ -245,6 +245,107 @@ func TestGetValidators(t *testing.T) {
+	assert.Equal(t, candidates[4].Address, validators[1].Address, "%v", validators)
+	assert.Equal(t, int64(0), validators[0].Height, "%v", validators)
+	assert.Equal(t, int64(0), validators[1].Height, "%v", validators)
+
+	// no change in voting power - no change in sort
+	ctx = ctx.WithBlockHeight(20)
+	keeper.setCandidate(ctx, candidates[4])
+	validators = keeper.GetValidators(ctx)
+	require.Equal(t, len(validators), n)
+	assert.Equal(t, candidates[3].Address, validators[0].Address, "%v", validators)
+	assert.Equal(t, candidates[4].Address, validators[1].Address, "%v", validators)
+
+	// change in voting power of both candidates, both still in v-set, no age change
+	candidates[3].Assets = sdk.NewRat(300)
+	candidates[4].Assets = sdk.NewRat(300)
+	keeper.setCandidate(ctx, candidates[3])
+	validators = keeper.GetValidators(ctx)
+	require.Equal(t, len(validators), n)
+	ctx = ctx.WithBlockHeight(30)
+	keeper.setCandidate(ctx, candidates[4])
+	validators = keeper.GetValidators(ctx)
+	require.Equal(t, len(validators), n, "%v", validators)
+	assert.Equal(t, candidates[3].Address, validators[0].Address, "%v", validators)
+	assert.Equal(t, candidates[4].Address, validators[1].Address, "%v", validators)
+
+	// now 2 max validators
+	params := keeper.GetParams(ctx)
+	params.MaxValidators = 2
+	keeper.setParams(ctx, params)
+	candidates[0].Assets = sdk.NewRat(500)
+	keeper.setCandidate(ctx, candidates[0])
+	validators = keeper.GetValidators(ctx)
+	require.Equal(t, uint16(len(validators)), params.MaxValidators)
+	require.Equal(t, candidates[0].Address, validators[0].Address, "%v", validators)
+	// candidate 3 was set before candidate 4
+	require.Equal(t, candidates[3].Address, validators[1].Address, "%v", validators)
+
+	// simulate scenario from https://github.com/cosmos/cosmos-sdk/issues/582#issuecomment-380757108
+	ctx = ctx.WithBlockHeight(40)
+	// candidate 4 kicked out temporarily
+	candidates[4].Assets = sdk.NewRat(200)
+	keeper.setCandidate(ctx, candidates[4])
+	validators = keeper.GetValidators(ctx)
+	require.Equal(t, uint16(len(validators)), params.MaxValidators)
+	require.Equal(t, candidates[0].Address, validators[0].Address, "%v", validators)
+	require.Equal(t, candidates[3].Address, validators[1].Address, "%v", validators)
+	// candidate 4 does not get spot back
+	candidates[4].Assets = sdk.NewRat(300)
+	keeper.setCandidate(ctx, candidates[4])
+	validators = keeper.GetValidators(ctx)
+	require.Equal(t, uint16(len(validators)), params.MaxValidators)
+	require.Equal(t, candidates[0].Address, validators[0].Address, "%v", validators)
+	require.Equal(t, candidates[3].Address, validators[1].Address, "%v", validators)
+	candidate, exists := keeper.GetCandidate(ctx, candidates[4].Address)
+	require.Equal(t, exists, true)
+	require.Equal(t, candidate.ValidatorHeight, int64(40))
+
+	// first transaction wins, https://github.com/cosmos/cosmos-sdk/issues/582#issuecomment-381250392
```

Comment:
should we add an actual explanation here rather than just referencing the github?

Reply:
Added an explanation, left the GH reference - I think linking to discussions (in addition to short in-code explanations) is useful, unless we're concerned about the code outlasting this GH repository?

Reply:
nah should be fine to leave the github link in there.. we can always remove later if we want right